Draft of refactoring tutorial + wistia embed component (#829)
dbt-labs/docs.getdbt.com: website/docs/tutorial/refactoring-legacy-sql.md

---
title: Refactoring legacy SQL to dbt SQL
id: refactoring-legacy-sql
description: This tutorial walks through refactoring a long SQL query (perhaps from a stored procedure) into modular dbt data models.
---

You may have already learned how to build dbt models from scratch. 

But in reality, you probably already have some queries or stored procedures that power analyses and dashboards, and now you’re wondering how to port those into dbt. 

> There are two parts to accomplish this: migration and refactoring. In this lesson we’re going to learn a process to help us turn legacy SQL code into modular dbt models.

Reviewer: ```suggestion
There are two parts to accomplish this: migration and refactoring. In this tutorial we’re going to learn a process to help us turn legacy SQL code into modular dbt models.
```

This step may sound simple, but if you're porting over an existing set of SQL transformations to a new SQL dialect, you will need to consider how your legacy SQL dialect differs from your new SQL flavor, and you may need to modify your legacy code to get it to run at all.  

This will commonly happen if you're migrating from a [stored procedure workflow on a legacy database](https://getdbt.com/analytics-engineering/case-for-elt-workflow/) into dbt + a cloud data warehouse.

Functions that you were using previously may not exist, or their syntax may shift slightly between SQL dialects. 

If you're not migrating data warehouses at the moment, then you can keep your SQL syntax the same. You have access to the exact same SQL dialect inside of dbt that you have querying directly from your warehouse.

## Create sources from table references

<WistiaVideo id="m1a5p32rny" />

To query from your data warehouse, we recommend creating [sources in dbt](/docs/building-a-dbt-project/using-sources) rather than querying the database table directly.

> This allows you to call the same table in multiple places with `{{ src('my_source', 'my_table') }}` rather than `my_database.my_schema.my_table`.

Reviewer: Curious while I'm here - does this work now? I missed an update 🤯

Reviewer 2: Tested `{{ src() }}` with 0.20.2 and it didn't work - so this just needs to be changed to `{{ source() }}` =)

We start here for several reasons:

#### Source freshness reporting
Using sources unlocks the ability to run [source freshness reporting](/docs/building-a-dbt-project/using-sources#snapshotting-source-data-freshness) to make sure your raw data isn't stale.

#### Easy dependency tracing
If you're migrating multiple stored procedures into dbt, with sources you can see which queries depend on the same raw tables. 

This allows you to consolidate modeling work on those base tables, rather than calling them separately in multiple places. 

<Lightbox src="/img/docs/building-a-dbt-project/sources-dag.png" title="Sources appear in green in your DAG in dbt docs" />

#### Build the habit of analytics-as-code
Sources are an easy way to get your feet wet using config files to define aspects of your transformation pipeline.

```yml
sources:
  - name: jaffle_shop
    tables:
      - name: orders
      - name: customers
```

> With a few lines of code in a `.yml` file in your dbt project's `/models` subfolder, you can now version control how you your data sources (Snowplow, Shopify, etc) map to actual database tables.

Reviewer: ```suggestion
With a few lines of code in a `.yml` file in your dbt project's `/models` subfolder, you can now version control how your data sources (Snowplow, Shopify, etc) map to actual database tables.
```

> For example, say you migrate from one [ETL tool](https://getdbt.com/analytics-engineering/etl-tools-a-love-letter/) to another, and the schema name on your tables changes from - source config files allow you to make that migration from a single file, and flip the change on with one pull request to your repo. 

Reviewer: ```suggestion
For example, say you migrate from one [ETL tool](https://getdbt.com/analytics-engineering/etl-tools-a-love-letter/) to another, and the schema name on your tables changes from - source config files allow you to make that migration from a single file, and flip the change on with one pull request to your repo. 
```
your tables changes from....you were on to something but I don't think it made its way in.

#### 1. Import CTEs

Let's start with our components, and identify raw data that is being used in our analysis. For this exercise, the components are three sources:

- jaffle_shop.customers
- jaffle_shop.orders
- stripe.payment

Let's make a CTE for each of these under the `Import CTEs` comment. These import CTEs should be only simple `select *` statements, but can have filters if necessary. 

We'll cover that later - for now, just use `select * from {{ source('schema', 'table') }}` for each, with the appropriate reference. Then, we will switch out all hard-coded references with our import CTE names.

#### 2. Logical CTEs

Logical CTEs contain unique transformations used to generate the final product, and we want to separate these into logical blocks. To identify our logical CTEs, we will follow subqueries in order. 

If a subquery has nested subqueries, we will want to continue moving down until we get to the first layer, then pull out the subqueries in order as CTEs, making our way back to the final select statement. 

> Name these CTEs as the alias that the subquery was given - you can rename it later, but for now it is best to make as little change as possible.

Reviewer: ```suggestion
Name these CTEs as the alias that the subquery was given - you can rename it later, but for now it is best to make as few changes as possible.
```
